numpy/core/src/umath/ufunc_object.c

```diff
@@ -1730,10 +1745,154 @@ make_arr_prep_args(npy_intp nin, PyObject *args, PyObject *kwds)
     }
 }
 
+/*
+ * Check whether any of the outputs of a gufunc has core dimensions.
+ */
+static int
+_has_output_coredims(PyUFuncObject *ufunc) {
+    int i;
+    for (i = ufunc->nin; i < ufunc->nin + ufunc->nout; ++i) {
+        if (ufunc->core_num_dims[i] > 0) {
+            return 1;
+        }
+    }
+    return 0;
+}
+
+/*
+ * Interpret a possible axes keyword argument, using it to remap the default
+ * axes (at the end of the operands) to new ones.
+ *
+ * Returns 0 on success, and -1 on failure
+ */
+static int
+_parse_axes_arg(PyUFuncObject *ufunc, PyObject *axes, PyArrayObject **op,
+                int broadcast_ndim, int **remap_axis) {
+    int nin = ufunc->nin;
+    int nout = ufunc->nout;
+    int nop = nin + nout;
+    int op_ndim, core_ndim, num_axes;
+    int i, j, iter_axis;
+    int requested_axis = 0;
+    int list_size = 0;
+    int item_is_tuple = 0;
+    PyObject *item = NULL;
+
+    if (!PyList_Check(axes)) {
+        PyErr_SetString(PyExc_TypeError, "axes should be a list.");
+        return -1;
+    }
+    list_size = PyList_Size(axes);
+    if (list_size != nop) {
+        if (list_size != nin || _has_output_coredims(ufunc)) {
+            PyErr_Format(PyExc_ValueError,
+                         "axes should be a list with an entry for all "
+                         "%d inputs and outputs; entries for outputs can only "
+                         "be omitted if none of them has core axes.",
+                         nop);
+            return -1;
+        }
+        for (i = nin; i < nop; i++) {
+            remap_axis[i] = NULL;
+        }
+    }
+    for (i = 0; i < list_size; ++i) {
+        int have_seen_axis[NPY_MAXDIMS] = {0};
+
+        core_ndim = ufunc->core_num_dims[i];
+        if (op[i] != NULL) {
+            op_ndim = PyArray_NDIM(op[i]);
+        }
+        else {
+            op_ndim = broadcast_ndim + core_ndim;
+        }
+        item = PyList_GetItem(axes, i);
+        item_is_tuple = PyTuple_Check(item);
+        if (item_is_tuple) {
+            num_axes = PyTuple_Size(item);
+        }
+        else {
+            /* try as integer */
+            requested_axis = (int)PyInt_AsLong(item);
```

Comment:
Does this allow `axis=1.0`?

Reply:
Indeed, it does. Now replaced by `PyNumber_AsSsize_t` (which goes through `__index__` if needed).